apps/web/src/lib/functions/file-loaders/epub/generate-epub-html.ts

```diff
@@ -139,32 +139,48 @@ export default function generateEpubHtml(
       htmlHref = itemIdToHtmlRef[itemIdRef];
     }
 
-    const regexResult = /.*<body(?:[^>]*id="(?<id>.+?)")*[^>]*>(?<body>(.|\s)+)<\/body>.*/.exec(
-      data[htmlHref] as string
-    )!;
+    let parsedContent = parser.parseFromString(data[htmlHref] as string, 'text/html');
+    let body = parsedContent.body;
 
-    const bodyId = regexResult?.groups?.id || '';
-    let innerHtml = regexResult?.groups?.body || '';
+    if (!body?.childElementCount) {
+      parsedContent = parser.parseFromString(data[htmlHref] as string, 'text/xml');
+      body = parsedContent.querySelector('body'); // XMLDocument doesn't seem to have the body property
+
+      if (!body?.childElementCount) {
+        throw new Error('Unable to find body tag while parsing EPUB content');
```

Comment:
a more precise error message would maybe be just something like "Unable to find valid body content"